File: csharp/mongodb-samples/Aggregation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MongoDB.Bson;
using MongoDB.Driver;
using NUnit.Framework;

namespace mongodb_samples
{
    [TestFixture]
    public class Aggregation
    {
        [Test]
        public void Test()
        {
            //creamos la conexion y seleccionamos la base de datos
            var server = MongoServer.Create("mongodb://localhost:27017");
            var database = server.GetDatabase("testdb");
            var collection = database.GetCollection<Person>("persons");
            collection.RemoveAll();

            var r = new Random();
            string[] names = { "Juan", "Antonio", "Pedro", "Maria", "Jordi", "Mario" };
            string[] apellidos = { "Gomez", "Perez" };

            for (int i = 1; i < 100; i++)
            {
                string apellido = apellidos[r.Next(0, 1)];
                //creamos un objeto persona
                var p = new Person { Age = r.Next(25, 55), Name = names[r.Next(0, 4)] + " " + apellido, Childs = new List<Child>(), Tags = new List<string>() {"A","B"}};
                p.Childs.Add(new Child { Name = names[r.Next(0, 4)] + " " + apellido, Age = r.Next(1, 12) });
                p.Childs.Add(new Child { Name = names[r.Next(0, 4)] + " " + apellido, Age = r.Next(1, 12) });

                //lo guardamos
                collection.Insert(p);
            }

            Console.WriteLine("Cuantas personas de cada Nombre hay");

            var operations = new[]{
                new BsonDocument {
                {
                    "$group", new BsonDocument{ { "_id" , "$Name" } , {"Total" , new BsonDocument{ {"$sum",1} } } }
                }
               }
            };

            var result = collection.Aggregate(operations);

            foreach (var document in result.ResultDocuments)
            {
                Console.WriteLine(document.ToJson());
            }

            Console.WriteLine();
            Console.WriteLine("Años Sumados de todas las personas x nombre");

            operations = new[]{
                new BsonDocument {
                {
                    "$group", new BsonDocument{ { "_id" , "$Name" } , {"Total" , new BsonDocument{ {"$sum","$Age"} } } }
                }
               }
            };

            result = collection.Aggregate(operations);

            foreach (var document in result.ResultDocuments)
            {
                Console.WriteLine(document.ToJson());
            }


            Console.WriteLine();
            Console.WriteLine("Cuantos hay y las edades unicas de los que se llaman Juan Gomez mayores de 40");

            string nameFilter = "Juan Gomez";
            
            operations = new[]{
                new BsonDocument{
                    {
                        "$match" , new BsonDocument
                                       {
                                           {"Name" , nameFilter }, 
                                           {"Age" , new BsonDocument{ {"$gte" , 40} } },
                                       }
                    } 
                }, 
                new BsonDocument {
                {
                    "$group", new BsonDocument{ { "_id" , "$Name" } , {"Total" , new BsonDocument{ {"$sum",1} } },
                              new BsonDocument("UniqueAges" , new BsonDocument( "$addToSet" , "$Age" )) }
                }               
               }               
            };

            result = collection.Aggregate(operations);

            foreach (var document in result.ResultDocuments)
            {
                Console.WriteLine(document.ToJson());
            }

        }
    }
}

File: python/aggregation.py
__author__ = 'emilio.torrens'

import pymongo
import random

#creamos la conexion y seleccionamos la base de datos
connection = pymongo.Connection('mongodb://localhost:27017')
database = connection['testdb']
database.persons.drop()

names = ["Juan", "Antonio", "Pedro", "Maria", "Jordi", "Mario"]
apellidos = ["Gomez","Perez"]

for id in range(1,100):
    #creamos un objeto persona
    apellido = apellidos[random.randrange(0,1)]
    person = dict(name= names[random.randrange(0,5)] + " " + apellido, age= random.randrange(25, 55), childs = [])
    person['childs'].append(dict(name= names[random.randrange(0,5)] + " " + apellido,age = random.randrange(1, 12)))
    person['childs'].append(dict(name= names[random.randrange(0,5)] + " " + apellido,age = random.randrange(1, 12)))

    #lo guardamos
    database.persons.insert(person)

print 'Cuantas personas de cada Nombre hay'

operation  = \
[
    {'$group': {
            '_id': '$name',
            'total': { '$sum': 1 } } }
]

results = database.command('aggregate', 'persons', pipeline= operation)

for result in results['result']:
    print result

print''
print 'Anyos Sumados de todas las personas x nombre'

operation  =\
[
    {'$group': {
        '_id': '$name',
        'total': { '$sum': '$age' } } }
]

results = database.command('aggregate', 'persons', pipeline= operation)

for result in results['result']:
    print result

print ''
print 'Cuantos hay y las edades unicas de los que se llaman Juan Gomez con edad >=40 y con hijos edad >= 3'

operation  =\
[
    {
        '$match': {'age':{'$gte': 35},'childs.age':{'$gte': 4}}
    },
    { '$group': {
        '_id': '$name',
        'total': { '$sum': 1 },
        'unique_ages': {'$addToSet': '$age'}} },
    { '$sort': { 'total': -1 } },
    { '$project' : { '_id' : 1, 'total' : 1, 'unique_ages' :1 } },
    #{ '$unwind' : '$unique_ages' }
]

results = database.command('aggregate', 'persons', pipeline= operation)

for result in results['result']:
    print result